### PR TITLE
Speed up hash->index conversion

### PR DESCRIPTION
**Pull request by Hackerpilot**
This results in a measurable speed increase in a data processing program that makes heavy use of hash tables.

This also fixes a bug in the (unlikely) case when `len` was 1.

### CODE REVIEW

**Comment by Hackerpilot** [comment]
This may have to wait... It seems that LDC doesn't support `tzcnt`.

File: src/containers/internal/hash.d [hunk 1]
```diff
@@ -51,10 +51,34 @@ size_t hashToIndex(const size_t hash, const size_t len) pure nothrow @nogc @safe
...
 	// It's amazing how much faster this makes the hash data structures
 	// when faced with low quality hash functions.
-	static if (size_t.sizeof == 8)
-		return (hash * 11_400_714_819_323_198_485UL) >>> (64 - bsr(len));
+	version (D_InlineAsm_X86_64)
+	{
+		asm @nogc
+		{
+			naked;
+			cmp RDI, 1;
```
Comment (Basile-z) [inline src/containers/internal/hash.d]: Take care to win64 , i think that register used are not the same (so RDI should be RCX maybe)

Comment (Hackerpilot) [reply src/containers/internal/hash.d]: Why did you have to remind me that Windows exists? I was hoping to forget.

[the remark this answers is not in the document]
Comment (Basile-z) [reply src/containers/internal/hash.d]: this have to be tested of course (i'm not sure).

**Comment by Hackerpilot** [comment]
> LDC will generate the same or better anyway.
Maybe I'm not using the right compiler switches, but in my builds I'm seeing that calls to some of the core.bitop functions are not even inlined.

I'm going to close this and see if I can come up with a better way to optimize this.

**Comment by Hackerpilot** [comment]
Even with `-O3 -mcpu=native` I'm seeing ldc generate code that says `callq  44d10 <_D4core5bitop3bsrFNaNbNiNfmZi>` instead of just running the desired cpu instructions. Looks like I'll need to manually optimize a few of the functions for the hash containers and unrolled list.

**Comment by Basile-z** [comment]
The problem with tznct is that it's not available with the core.bitop automagic, which means no inlining possible. I [added it](https://github.com/dlang/dmd/pull/6364) a couple of years ago to iasm, planned to add more BMI2 instructions but never continued this project at last. At some point [people asked to have them via druntime](https://github.com/dlang/dmd/pull/6364#issuecomment-269303051), but anyway, you have no other options nowadays than inline asm.

**Comment by Hackerpilot** [comment]
`ldc2 -singleobj -flto=full -O3 -mcpu=native` still generates a `call` instead of a `bsr` on a broadwell machine.

**Comment by Hackerpilot** [comment]
Apparently I'm dumb and the solution is `import ldc.intrinsics : llvm_cttz;`

**Pull request closed without merging**